gopherage/cmd/metadata/metadata.go

```diff
@@ -0,0 +1,141 @@
+	base_ref   = "PULL_BASE_REF"
+	repo_owner = "REPO_OWNER"
+)
+
+var (
+	current_commit = []string{"rev-parse", "HEAD"}
+)
+
+type flags struct {
+	outputFile    string
+	host          string
+	project       string
+	commitID      string
+	ref           string
+	workspaceRoot string
+}
+
+type CoverageMetadata struct {
+	Host     string `json:"host"`
+	Project  string `json:"project"`
+	Root     string `json:"workspace_root"`
+	Ref      string `json:"ref"`
+	CommitID string `json:"commit_id"`
+}
+
+// MakeCommand returns a `junit` command.
+func MakeCommand() *cobra.Command {
+	flags := &flags{}
+	cmd := &cobra.Command{
+		Use:   "metadata [...fields]",
+		Short: "Produce json file containing metadata about coverage collection.",
+		Long:  `Builds a json file containing information about the repo .`,
+		Run: func(cmd *cobra.Command, args []string) {
+			run(flags, cmd, args)
+		},
+	}
+	cmd.Flags().StringVarP(&flags.outputFile, "output", "o", "-", "output file")
+	cmd.Flags().StringVarP(&flags.host, "host", "h", "", "Name of repo host")
+	cmd.Flags().StringVarP(&flags.project, "project", "p", "", "Project name")
+	cmd.Flags().StringVarP(&flags.commitID, "commit", "c", "", "Current Commit Hash (git rev-parse HEAD)")
+	cmd.Flags().StringVarP(&flags.ref, "ref", "r", "", "Current branch ref (git branch --show-current).")
+	cmd.Flags().StringVarP(&flags.workspaceRoot, "root", "w", "", "path to root of repo")
+	return cmd
+}
+
+func run(flags *flags, cmd *cobra.Command, args []string) {
+	if flags.project == "" {
+		ref := os.Getenv(repo_owner)
+		if ref == "" {
+			fmt.Fprintf(os.Stdout, "Failed to collect project from ENV: (%s) not found", repo_owner)
+			cmd.Usage()
+			os.Exit(1)
+		}
+		flags.ref = ref
+	}
+
+	if flags.commitID == "" {
+		cmd := exec.Command("git", current_commit...)
+		var out bytes.Buffer
+		cmd.Stdout = &out
+
+		err := cmd.Run()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to fetch Commit Hash run from within covered repo: %v.", err)
+			os.Exit(1)
+		}
+		flags.commitID = strings.TrimSuffix(out.String(), "\n")
+	}
+
+	if flags.ref == "" {
+		ref := os.Getenv(base_ref)
+		if ref == "" {
+			fmt.Fprintf(os.Stdout, "Failed to collect ref from ENV: (%s) not found, using 'HEAD'", base_ref)
+			flags.ref = "HEAD"
+		}
+		flags.ref = ref
+	}
+
+	metadata := &CoverageMetadata{
+		Host:     flags.host,
+		Project:  flags.project,
+		Root:     flags.workspaceRoot,
+		Ref:      flags.ref,
+		CommitID: flags.commitID,
+	}
+
+	j, err := json.Marshal(metadata)
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to build json: %v.", err)
+		os.Exit(1)
+	}
+
+	var file io.WriteCloser
+	if flags.outputFile == "-" {
+		file = os.Stdout
+	} else {
+		file, err = os.Create(flags.outputFile)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to create file: %v.", err)
+			os.Exit(1)
+		}
+	}
+	if _, err := file.Write(j); err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to write json: %v.", err)
+		os.Exit(1)
```

Comment:
in future, we should refactor this to use runE and handle errors more gracefully instead of os.exit everywhere.

Reply:
I used the pattern I saw in other sub commands. I will fix this across the board at once.